### PR TITLE
ContinuousLocking4Reputation scheme

### PR DESCRIPTION
**Pull request by orenyodfat**
bootstrap scheme for nectar dao.

### CODE REVIEW

File: contracts/schemes/ContinuousLockingToken4Reputation.sol [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+     * @param _agreementHash is a hash of agreement required to be added to the TX by participants
+     */
+    function initialize(
+        Avatar _avatar,
+        uint256 _reputationReward,
+        uint256 _startTime,
+        uint256 _periodsUnit,
+        uint256 _redeemEnableTime,
+        uint256 _maxLockingPeriods,
+        uint256 _repRewardConstA,
```
Comment (leviadam) [inline contracts/schemes/ContinuousLockingToken4Reputation.sol]: No @param comment.

File: contracts/schemes/ContinuousLockingToken4Reputation.sol [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    event Release(bytes32 indexed _lockingId, address indexed _beneficiary, uint256 _amount);
+    event LockToken(address indexed _locker, bytes32 indexed _lockingId, uint256 _amount, uint256 _period);
+
+    struct Locking {
+        uint256 totalScore;
+        // A mapping from locker addresses to their locking score.
+        mapping(address=>uint256) scores;
+    }
+
+    struct Locker {
```
Comment (leviadam) [inline contracts/schemes/ContinuousLockingToken4Reputation.sol]: This is not representing a locker but rather a specific locking.

File: contracts/schemes/ContinuousLockingToken4Reputation.sol [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        require(_period > 0, "locking period equal to zero");
+        require((_lockingPeriodToLockIn + _period) <= MAX_PERIODS, "exceed max allowed periods");
+        address(token).safeTransferFrom(msg.sender, address(this), _amount);
+        // solhint-disable-next-line not-rely-on-time
+        uint256 lockingPeriodToLockIn = (now - startTime) / periodsUnit;
+        require(lockingPeriodToLockIn == _lockingPeriodToLockIn, "locking is not active");
+        uint256 j = _period;
+        //fill in the next lockings scores.
+        //todo : check limitation of _period and require that on the init function.
+        for (int256 i = int256(lockingPeriodToLockIn + _period - 1); i >= int256(lockingPeriodToLockIn); i--) {
```
Comment (leviadam) [inline contracts/schemes/ContinuousLockingToken4Reputation.sol]: Isn't it simpler to write the loop forward?

**Comment by orenyodfat** [comment]
thanks @leviadam . 
see 00e2858 and b22f5c3

File: contracts/schemes/ContinuousLockingToken4Reputation.sol [hunk 4]
```diff
@@ -21,45 +21,47 @@ contract ContinuousLocking4Reputation is Agreement {
...
+    struct Lock {
         uint256 amount;
         uint256 lockingTime;
         uint256 period;
     }
 
     // A mapping from lockers addresses their lock balances.
-    mapping(address => mapping(bytes32=>Locker)) public lockers;
+    mapping(address => mapping(bytes32=>Lock)) public lockers;
     // A mapping from locking index to locking.
```
Comment (leviadam) [inline contracts/schemes/ContinuousLockingToken4Reputation.sol]: A mapping from batch index to batch.

File: contracts/schemes/ContinuousLockingToken4Reputation.sol [hunk 5]
```diff
@@ -21,45 +21,47 @@ contract ContinuousLocking4Reputation is Agreement {
...
+    mapping(uint256=>Batch) public batches;
 
     Avatar public avatar;
     uint256 public reputationRewardLeft;
     uint256 public startTime;
     uint256 public numberOfLockingPeriods;
     uint256 public redeemEnableTime;
-    uint256 public maxLockingPeriods;
+    uint256 public maxLockingBatches;
     uint256 public periodsUnit;
```
Comment (leviadam) [inline contracts/schemes/ContinuousLockingToken4Reputation.sol]: batchTime?

File: contracts/schemes/ContinuousLockingToken4Reputation.sol [hunk 6]
```diff
@@ -117,13 +128,13 @@ contract ContinuousLocking4Reputation is Agreement {
     function redeem(address _beneficiary, bytes32 _lockingId) public returns(uint256 reputation) {
         // solhint-disable-next-line not-rely-on-time
         require(now > redeemEnableTime, "now > redeemEnableTime");
-        Locker storage locker = lockers[_beneficiary][_lockingId];
+        Lock storage locker = lockers[_beneficiary][_lockingId];
         uint256 periodToRedeemFrom = (locker.lockingTime - startTime) / periodsUnit;
         // solhint-disable-next-line not-rely-on-time
         uint256 currentLockingPeriod = (now - startTime) / periodsUnit;
```
Comment (leviadam) [inline contracts/schemes/ContinuousLockingToken4Reputation.sol]: currentBatch?

**Comment by orenyodfat** [comment]
thanks @leviadam

**Pull request merged**